extensions/mssql/src/main.ts

```diff
@@ -0,0 +1,131 @@
+
+	let clientOptions: ClientOptions = {
+		providerId: Constants.providerId,
+		errorHandler: new LanguageClientErrorHandler(),
+		serverConnectionMetadata: undefined
+	};
+
+	const installationStart = Date.now();
+	serverdownloader.getOrDownloadServer().then(e => {
+		const installationComplete = Date.now();
+		let serverOptions = generateServerOptions(e);
+		languageClient = new SqlOpsDataClient(Constants.serviceName, serverOptions, clientOptions);
+		const processStart = Date.now();
+		languageClient.onReady().then(() => {
+			const processEnd = Date.now();
+			statusView.text = 'Service Started';
+			setTimeout(() => {
+				statusView.hide();
+			}, 1500);
+			Telemetry.sendTelemetryEvent('startup/LanguageClientStarted', {
+				installationTime: String(installationComplete - installationStart),
+				processStartupTime: String(processEnd - processStart),
+				totalTime: String(processEnd - installationStart),
+				beginningTimestamp: String(installationStart)
+			});
+		});
+		statusView.show();
+		statusView.text = 'Starting service';
+		languageClient.start();
+		credentialsStore.start();
+		resourceProvider.start();
+	}, e => {
+		Telemetry.sendTelemetryEvent('ServiceInitializingFailed');
+		vscode.window.showErrorMessage('Failed to start Sql tools service');
+	});
+
+	let contextProvider = new ContextProvider();
+	context.subscriptions.push(contextProvider);
+	context.subscriptions.push(credentialsStore);
+	context.subscriptions.push(resourceProvider);
+	context.subscriptions.push({ dispose: () => languageClient.stop() });
+}
+
+function generateServerOptions(executablePath: string): ServerOptions {
+	let launchArgs = [];
+	launchArgs.push('--log-dir');
+	let logFileLocation = path.join(Utils.getDefaultLogLocation(), 'mssql');
+	launchArgs.push(logFileLocation);
+
+	return { command: executablePath, args: launchArgs, transport: TransportKind.stdio };
+}
+
+function generateHandleServerProviderEvent() {
+	let dots = 0;
+	return (e: string, ...args: any[]) => {
+		outputChannel.show();
+		statusView.show();
+		switch (e) {
+			case Events.INSTALL_START:
+				outputChannel.appendLine(`Installing ${Constants.serviceName} service to ${args[0]}`);
+				statusView.text = 'Installing Service';
+				break;
+			case Events.INSTALL_END:
+				outputChannel.appendLine('Installed');
+				break;
+			case Events.DOWNLOAD_START:
+				outputChannel.appendLine(`Downloading ${args[0]}`);
+				outputChannel.append(`(${Math.ceil(args[1] / 1024)} KB)`);
+				statusView.text = 'Downloading Service';
+				break;
+			case Events.DOWNLOAD_PROGRESS:
+				let newDots = Math.ceil(args[0] / 5);
+				if (newDots > dots) {
+					outputChannel.append('.'.repeat(newDots - dots));
+					dots = newDots;
+				}
+				break;
+			case Events.DOWNLOAD_END:
+				outputChannel.appendLine('Done!');
+				break;
+		}
+	};
+}
+
+// this method is called when your extension is deactivated
+export function deactivate(): void {
```

Comment:
https://github.com/Microsoft/sqlopsstudio/pull/909/files/b2419de5994ffaed1e53f48c4f0adc9d0ce6d1c1#diff-6558b65a4d6dd3857b2170e2a27cb3cfR85

The client will be shutdown when the extension is disposed. I think that should cover what you are saying here.

Reply:
Hmm. I see what you're saying, that's an interesting way of doing it. Only thing I might add is that I think we would ideally also run the same code as disposal in the deactivated. I'm not 100% on the scenarios when deactivate is run (my original thought is that this is when the extension is disabled without restarting, which I suppose isn't possible since mssql is baked into SQLOps).